File: signalpilot/web/lib/deliverable-theme.ts
import type { CSSProperties } from "react";
import type { DeliverableTheme } from "./types";

// Keep in sync with gateway/gateway/models/deliverable_theme.py.
export const DEFAULT_POSITIVE_COLOR = "#00ff88";
export const DEFAULT_CHART_SERIES = ["#087a3d", "#0fa45a", "#35c978", "#8eeaa8", "#c8f8d4", "#e7fcec"];

export const DEFAULT_DELIVERABLE_THEME: DeliverableTheme = {
  version: 1,
  colors: {
    bg: "#050505",
    surface: "#0a0a0a",
    surface_alt: "#0d0d0d",
    border: "#222222",
    text: "#eeeeee",
    muted: "#999999",
    accent: "#ffffff",
    positive: DEFAULT_POSITIVE_COLOR,
    warning: "#ffaa00",
    negative: "#ff4444",
    chart_grid: "#1f1f1f",
    chart_axis: "#999999",
  },
  chart_series: DEFAULT_CHART_SERIES,
  font_family: '"SF Mono", "Fira Code", "Cascadia Code", ui-monospace, monospace',
  font_size_base_px: 14,
  spacing_unit_px: 8,
  radius_px: 6,
};

type ThemeCssVars = CSSProperties & Record<`--sp-${string}`, string>;

const HEX_RE = /^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/;

function expandHex(value: string): string {
  if (/^#[0-9a-fA-F]{3}$/.test(value)) {
    return `#${value[1]}${value[1]}${value[2]}${value[2]}${value[3]}${value[3]}`.toLowerCase();
  }
  return value.toLowerCase();
}

function mixChannel(channel: number, target: number, amount: number): number {
  return Math.round(channel + (target - channel) * amount);
}

function mixRgb(rgb: [number, number, number], target: [number, number, number], amount: number): string {
  return `#${rgb
    .map((channel, index) => mixChannel(channel, target[index], amount).toString(16).padStart(2, "0"))
    .join("")}`;
}

export function chartSeriesFromPositive(positive: string): string[] {
  if (!HEX_RE.test(positive)) return [...DEFAULT_CHART_SERIES];
  const normalized = expandHex(positive);
  if (normalized === DEFAULT_POSITIVE_COLOR) return [...DEFAULT_CHART_SERIES];

  const rgb: [number, number, number] = [
    Number.parseInt(normalized.slice(1, 3), 16),
    Number.parseInt(normalized.slice(3, 5), 16),
    Number.parseInt(normalized.slice(5, 7), 16),
  ];
  const black: [number, number, number] = [0, 0, 0];
  const white: [number, number, number] = [255, 255, 255];
  return [
    mixRgb(rgb, black, 0.35),
    mixRgb(rgb, black, 0.18),
    normalized,
    mixRgb(rgb, white, 0.32),
    mixRgb(rgb, white, 0.58),
    mixRgb(rgb, white, 0.78),
  ];
}

export function themeWithGeneratedChartSeries(theme: DeliverableTheme): DeliverableTheme {
  return {
    ...theme,
    colors: { ...theme.colors },
    chart_series: chartSeriesFromPositive(theme.colors.positive),
  };
}

export function themeToCssVars(theme: DeliverableTheme): ThemeCssVars {
  const chartSeries = chartSeriesFromPositive(theme.colors.positive);
  const vars: ThemeCssVars = {
    "--sp-bg": theme.colors.bg,
    "--sp-surface": theme.colors.surface,
    "--sp-surface-alt": theme.colors.surface_alt,
    "--sp-border": theme.colors.border,
    "--sp-text": theme.colors.text,
    "--sp-muted": theme.colors.muted,
    "--sp-accent": theme.colors.accent,
    "--sp-positive": theme.colors.positive,
    "--sp-warning": theme.colors.warning,
    "--sp-negative": theme.colors.negative,
    "--sp-chart-grid": theme.colors.chart_grid,
    "--sp-chart-axis": theme.colors.chart_axis,
    "--sp-font": theme.font_family,
    "--sp-font-size": `${theme.font_size_base_px}px`,
    "--sp-space": `${theme.spacing_unit_px}px`,
    "--sp-space-2": `${theme.spacing_unit_px * 2}px`,
    "--sp-space-3": `${theme.spacing_unit_px * 3}px`,
    "--sp-space-4": `${theme.spacing_unit_px * 4}px`,
    "--sp-radius": `${theme.radius_px}px`,
  };
  chartSeries.forEach((color, index) => {
    vars[`--sp-chart-${index + 1}`] = color;
  });
  return vars;
}

export function cloneTheme(theme: DeliverableTheme): DeliverableTheme {
  return {
    ...theme,
    colors: { ...theme.colors },
    chart_series: [...theme.chart_series],
  };
}

File: signalpilot/gateway/gateway/models/deliverable_theme.py
"""Theme tokens for generated SignalPilot HTML deliverables."""

from __future__ import annotations

import re

from pydantic import BaseModel, Field, field_validator

HEX_COLOR_RE = re.compile(r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")
DEFAULT_CHART_SERIES = ["#087a3d", "#0fa45a", "#35c978", "#8eeaa8", "#c8f8d4", "#e7fcec"]
DEFAULT_FONT_FAMILY = '"SF Mono", "Fira Code", "Cascadia Code", ui-monospace, monospace'


def _validate_hex_color(value: str, field_name: str) -> str:
    if not HEX_COLOR_RE.fullmatch(value):
        raise ValueError(f"{field_name} must be a #rgb or #rrggbb hex color")
    return value.lower()


class ThemeColors(BaseModel):
    bg: str = "#050505"
    surface: str = "#0a0a0a"
    surface_alt: str = "#0d0d0d"
    border: str = "#222222"
    text: str = "#eeeeee"
    muted: str = "#999999"
    accent: str = "#ffffff"
    positive: str = "#00ff88"
    warning: str = "#ffaa00"
    negative: str = "#ff4444"
    chart_grid: str = "#1f1f1f"
    chart_axis: str = "#999999"

    @field_validator("*")
    @classmethod
    def validate_hex_colors(cls, value: str, info) -> str:
        return _validate_hex_color(value, info.field_name)


class DeliverableTheme(BaseModel):
    version: int = 1
    colors: ThemeColors = Field(default_factory=ThemeColors)
    chart_series: list[str] = Field(default_factory=lambda: DEFAULT_CHART_SERIES.copy(), min_length=3, max_length=8)
    font_family: str = Field(default=DEFAULT_FONT_FAMILY, min_length=1, max_length=240)
    font_size_base_px: int = Field(default=14, ge=10, le=24)
    spacing_unit_px: int = Field(default=8, ge=4, le=24)
    radius_px: int = Field(default=6, ge=0, le=24)

    @field_validator("chart_series")
    @classmethod
    def validate_chart_series(cls, value: list[str]) -> list[str]:
        return [_validate_hex_color(color, "chart_series") for color in value]

    @field_validator("font_family")
    @classmethod
    def validate_font_family(cls, value: str) -> str:
        cleaned = value.strip()
        if any(char in cleaned for char in "{};<>"):
            raise ValueError("font_family contains unsupported CSS control characters")
        return cleaned
